packages/react-reconciler/src/ReactFiberHooks.js

```diff
@@ -168,6 +184,10 @@ let renderExpirationTime: ExpirationTime = NoWork;
 // the work-in-progress hook.
 let currentlyRenderingFiber: Fiber = (null: any);
 
+// The work-in-progress root fiber.
+// Used by mutable source hook to detect tears from pending updates.
+let currentlyRenderingRoot: FiberRoot = (null: any);
```

Comment:
The other functions are able to be executed mostly standalone and operate on a single Fiber. Passing the argument alone ensures that we do that.

I've been thinking about using a few synchronous passes a couple of times that don't change the workInProgressFiber in the work loop but instead calls begin/complete in its own loop. E.g. for computing exit transition states with different contexts. It would be nice to preserve that property.

With resuming or even smaller similar optimizations (like reusing completed suspended trees) it'll be very important that we don't accidentally use global state since a render of a subtree from one pass needs to be able to be reused in another. I think for gesture calculation that will also be a problem.

This whole approach breaks that rule but that's inherent to this special problem space and not general. We think that we can find a suitable workaround for this one special case.

However, in general we should never operate on the root in begin/complete. That's why I don't think we should be passing the root down through begin since it'll encourage us to use it in more places but this is the one legit case and even this one is on the edge.

Passing lots of unused variables on the stack also has perf implications.

Reply:
I didn't like passing it in all of the places either, but didn't feel comfortable adding a getter for it. (That seems to encourage other places to use it more than explicitly passing.)

I *think* Sebastian is agreeing with Andrew's suggestion but I'm not entirely sure because of:
> However, in general we should never operate on the root in begin/complete.

Pending a decision on [this comment](https://github.com/facebook/react/pull/18000#discussion_r379257652) (which I'm still thinking about), I need to access the current root.

Reply:
> Pending a decision on this comment (which I'm still thinking about), I need to access the current root.

To clarify, you'd have to access the root in that proposal, too. It would be `root.lastPendingMutationTime`.